libbeat/cfgfile/reload.go

```diff
@@ -137,7 +137,10 @@ func (rl *Reloader) Check(runnerFactory RunnerFactory) error {
 		if !c.Config.Enabled() {
 			continue
 		}
-		_, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		runner, err := runnerFactory.Create(rl.pipeline, c.Config, c.Meta)
+		if runner != nil {
+			runner.Stop()
```

Comment:
I went through the different Create and Stop functions of the runners. I found a potential issue here in Metricbeat: https://github.com/elastic/beats/blob/6.7/metricbeat/mb/module/runner.go#L92

When the runner is stopped, it removes the module from the module List. As Start is never called, it means the entry was never added to the list. We have a similar issue on filebeat side: https://github.com/elastic/beats/blob/master/filebeat/fileset/factory.go#L176

Perhaps the best short term fix is to follow the other suggestion @andrewvc had to start using `CheckConfig` here but also pass pipeline and metadata, so in FB and MB the method can call Create?

Reply:
@ruflin I agree. Unfortunately, after starting work on that refactor I found that autodiscover uses `CheckConfig` and has no pipeline instance to pass it here: https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L193

I propose we pause here and work on the real fix in https://github.com/elastic/beats/issues/11313

The hacks are piled too high here, and we should just do the right thing even if it's painful.

Reply:
++